x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/LookupJoinTypesIT.java

```diff
@@ -0,0 +1,463 @@
+            case BOOLEAN -> true;
+            case DATETIME, DATE_NANOS -> "2025-04-02T12:00:00.000Z";
+            case IP -> "127.0.0.1";
+            case KEYWORD, TEXT -> "key";
+            case BYTE, SHORT, INTEGER -> 1;
+            case LONG -> 1L;
+            case HALF_FLOAT, FLOAT, DOUBLE -> 1.0;
+            default -> throw new IllegalArgumentException("Unsupported type: " + type);
+        };
+    }
+
+    private static class TestConfigs {
+        final String group;
+        final Map<String, TestConfig> configs;
+
+        TestConfigs(String group) {
+            this.group = group;
+            this.configs = new LinkedHashMap<>();
+        }
+
+        private boolean exists(String indexName) {
+            return configs.containsKey(indexName);
+        }
+
+        private void add(TestConfig config) {
+            if (configs.containsKey(config.indexName())) {
+                throw new IllegalArgumentException("Duplicate index name: " + config.indexName());
+            }
+            configs.put(config.indexName(), config);
+        }
+
+        private void addPasses(DataType mainType, DataType lookupType) {
+            add(new TestConfigPasses(mainType, lookupType, true));
+        }
+
+        private void addEmptyResult(DataType mainType, DataType lookupType) {
+            add(new TestConfigPasses(mainType, lookupType, false));
+        }
+
+        private void addFails(DataType mainType, DataType lookupType) {
+            String fieldName = "field_" + mainType.esType();
+            String errorMessage = String.format(
+                Locale.ROOT,
+                "JOIN left field [%s] of type [%s] is incompatible with right field [%s] of type [%s]",
+                fieldName,
+                mainType.widenSmallNumeric(),
+                fieldName,
+                lookupType.widenSmallNumeric()
+            );
+            add(
+                new TestConfigFails<>(
+                    mainType,
+                    lookupType,
+                    VerificationException.class,
+                    e -> assertThat(e.getMessage(), containsString(errorMessage))
+                )
+            );
+        }
+
+        private void addFailsText(DataType mainType) {
+            String fieldName = "field_" + mainType.esType();
+            String errorMessage = String.format(Locale.ROOT, "JOIN with right field [%s] of type [TEXT] is not supported", fieldName);
+            add(
+                new TestConfigFails<>(
+                    mainType,
+                    DataType.TEXT,
+                    VerificationException.class,
+                    e -> assertThat(e.getMessage(), containsString(errorMessage))
+                )
+            );
+        }
+
+        private <E extends Exception> void addFails(DataType mainType, DataType lookupType, Class<E> exception, Consumer<E> assertion) {
+            add(new TestConfigFails<>(mainType, lookupType, exception, assertion));
+        }
```

Comment:
I think this was meant to be used in the other `addFails` methods above? Currently, I think it is unused.